src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java

```diff
@@ -418,6 +418,11 @@ private int read() {
                 nextBuffer.position(0);
                 return n;
             } catch (IOException ex) {
+                need2Read = false;
+                haveNext = false;
```

Comment:
This shouldn't be required - `need2Read`/`haveNext` will be set by `hasNext()`; I'd prefer if we kept the logic there.

Reply:
We can close the stream in `hasNext()`, but we need to move `close()` from `read()`.
`need2Read` and `haveNext` will be set to `false` if `read()` returns -1. However `read()` returns -1 when IOE or EOF happen. Is it ok? I concern to change location of `close()` - it means the change of side-effect.

Reply:
You misunderstand. `read()` should call `close();` when `read` returns `-1`, `hasNext()` will take care of updating `need2Read` and `haveNext`;